classes/abstracts/ActionScheduler_Abstract_QueueRunner.php

```diff
@@ -86,13 +86,94 @@ public function process_action( $action_id, $context = '' ) {
 	 * @param int $action_id
 	 */
 	protected function schedule_next_instance( ActionScheduler_Action $action, $action_id ) {
+		// If a recurring action has been consistently failing, we may wish to stop rescheduling it.
+		if (
+			ActionScheduler_Store::STATUS_FAILED === $this->store->get_status( $action_id )
+			&& $this->recurring_action_is_consistently_failing( $action, $action_id )
+		) {
+			ActionScheduler_Logger::instance()->log(
+				$action_id,
+				__( 'This action appears to be consistently failing. A new instance will not be scheduled.', 'action-scheduler' )
+			);
+
+			return;
+		}
+
 		try {
 			ActionScheduler::factory()->repeat( $action );
 		} catch ( Exception $e ) {
 			do_action( 'action_scheduler_failed_to_schedule_next_instance', $action_id, $e, $action );
 		}
 	}
 
+	/**
+	 * Determine if the specified recurring action has been consistently failing.
+	 *
+	 * @param ActionScheduler_Action $action    The recurring action to be rescheduled.
+	 * @param int                    $action_id The ID of the recurring action.
+	 *
+	 * @return bool
+	 */
+	private function recurring_action_is_consistently_failing( ActionScheduler_Action $action, $action_id ) {
+		// Fetch the most recent historic actions having the same hook and argument set.
```

Comment:
May I suggest a bit more performant algorithm to solve this method:
1. Get the last `n`th action ID that failed for the same hook and args, where n is `action_scheduler_recurring_action_failure_threshold` using `per_page = 1`, `status = failed` and `offset = n`.  If there isn't any then`return false`. Store this action ID somewhere.
2. Get the last `n`th action ID with the same hook and args, but this time don't pass the `status = failed` filter. If the last n actions have been consistently failing, then both these ID have to be the same. So return true if this ID is the same as the previous one otherwise return false.

This will be more performant than the current query because you can do the comparison in just 2 queries, instead of `n+1` queries like we do now (and hopefully bit less code). This won't take schedule into account, but I think that's fine as we use the method in deciding whether to schedule another recurring action or not (so we know that the current action is recurring for sure).

Reply:
@vedanshujain—that's a great idea, thank you.

> This won't take schedule into account, but I think that's fine as we use the method in deciding whether to schedule another recurring action or not (so we know that the current action is recurring for sure).

My only nit would be that, with this strategy, we don't know if all of those failing actions were recurring—we will only know that they have the same hook (ie, this algorithm does not account for a mix of single and recurring actions using the same hook).

Even so, it is probably still a very reasonable way to go. Talking it out some more:

- The specifics of the schedule (beyond the fact we're primarily interested in recurring actions) probably don't matter all that much: after all, when an action is triggered, the callback doesn't know anything about the schedule—it just knows what its arguments are.
- Taking this a step further, I am now wondering if checking that the args match is actually the best policy, because it means we will fail to detect and halt a failing sequence like the following:

| Hook | Args | Status |
| --- | --- | --- |
| foo | [ 'page' => 1 ] | failed |
| foo | [ 'page' => 2 ] | failed |
| foo | [ 'page' => 3 ] | failed |
| ... | ... | ... |

To that end, I think I will:

- Adopt the strategy you proposed
- Perform the matching based only on the hook and not the args

Let me know if you have thoughts on this!

Reply:
Updated in https://github.com/woocommerce/action-scheduler/pull/806/commits/75b2766085b614d79699acfaacd29a1617cc1943